### PR TITLE
Approved machine state

### PR DESCRIPTION
**Pull request by gmethvin**
This makes it so users who download the extension without having logged in on that machine are directed to the kifi.com site. If they can log in on the site it will trigger the extension to reload.

Any time we have either an approved user cookie on the machine or an active kifi extension installation in local storage, we'll authenticate with facebook as usual.

### CODE REVIEW

**Comment by 2is10** [comment]
Hmm... This pull request duplicates some code in the extension. Also, I'm not sure we need the new content script. I'd like to understand the high-level goal(s) and the urgency. I'm pretty sure we're already planning more general-purpose session management for next week's run. Let's discuss in the morning.

**Comment by gmethvin** [comment]
There is no particular urgency. Eishay talked about this with me last week and it was marked as a P2 for this week's run. I'm happy to wait on merging this.

**Comment by eishay** [comment]
can we either merge or close this pull request (keep the branch around)?

File: shoebox/app/com/keepit/controllers/website/HomeController.scala [hunk 1]
```diff
@@ -22,25 +20,28 @@ class HomeController @Inject() (db: Database,
...
+        }
 
-      val userCanInvite = (request.experiments & Set(ExperimentTypes.ADMIN, ExperimentTypes.CAN_INVITE)).nonEmpty
+        val userCanInvite = (request.experiments & Set(ExperimentTypes.ADMIN, ExperimentTypes.CAN_INVITE)).nonEmpty
 
-      Ok(views.html.website.userHome(request.user, friendsOnKifi, userCanInvite))
-    }
+        Ok(views.html.website.userHome(request.user, friendsOnKifi, userCanInvite))
+          .withCookies(userStateCookie.encodeAsCookie(Some(request.user.state)))
+    }).withCookies(userStateCookie.encodeAsCookie(Some(request.user.state)))
```
Comment (2is10) [inline shoebox/app/com/keepit/controllers/website/HomeController.scala]: I'd prefer we don't keep the user state in a cookie (sent with every request).

File: packrat/scripts/kifi_site.js [hunk 2]
```diff
@@ -0,0 +1,8 @@
+// @match /^https?:\/\/(dev\.ezkeep\.com:9000|(www\.)?(kifi|keepitfindit)\.com)/
+// @require scripts/api.js
+
+!(function () {
+  if (document.getElementsByClassName("kifi-logged-in").length) {
```
Comment (2is10) [inline packrat/scripts/kifi_site.js]: I just search the shoebox codebase for `kifi-logged-in` but found no match. What is this supposed supposed to find? On what page(s)? Could the match expression be more specific?

File: packrat/scripts/kifi_site.js [hunk 3]
```diff
@@ -0,0 +1,8 @@
...
+    api.port.emit("load_extension");
```
Comment (2is10) [inline packrat/scripts/kifi_site.js]: How about calling this message something more specific? Perhaps `"authentication_detected"`. Other meanings could be inferred from `"load_extension"`, such as `chrome.runtime.reload()`.

File: packrat/main.js [hunk 4]
```diff
@@ -591,6 +591,16 @@ api.port.on({
...
   add_deep_link_listener: function(locator, _, tab) {
     createDeepLinkListener(locator, tab.id);
+  },
+  load_extension: function() {
+    if (isApproved) return;
+    checkApproved(function (approved) {
+      if (!approved) return;
+      authenticate(function() {
+        api.log("[main] authenticated");
+        api.tabs.eachSelected(subscribe);
```
Comment (2is10) [inline packrat/main.js]: How about reusing the existing `authenticate` callback that does this stuff, plus taking the user to `/getting-started` if appropriate?

File: packrat/scripts/kifi_site.js (hunk 2 above)
Comment (gmethvin) [inline packrat/scripts/kifi_site.js]: I'm checking for a specific class on the home page you reach after you log in (included in this pull request).

Comment (2is10) [inline packrat/scripts/kifi_site.js]: I see, thanks. As long as that page's URL requires authentication, we shouldn't need a content script at all. We can just check for its URL in `api.tabs.on.loading` or `api.tabs.on.ready` in main.js. (Those don't get invoked in the case of a redirect.)

File: shoebox/app/com/keepit/controllers/website/HomeController.scala (hunk 1 above)
Comment (gmethvin) [inline shoebox/app/com/keepit/controllers/website/HomeController.scala]: I just need some way to set this state and access it from within the extension. What would be a good alternative?

**Comment by gmethvin** [comment]
Okay. I basically completely rewrote this based on the discussion @2is10 and I had today.

File: packrat/main.js [hunk 5]
```diff
@@ -686,6 +686,13 @@ function createDeepLinkListener(locator, tabId) {
   });
 }
 
+api.tabs.on.ready.add(function kifiLoginListener(tab) {
+  // check to see if this is the home page, if so try authenticating again if we don't have an installation
+  if (tab.url.replace(/\/#.*$/, "") === webBaseUri() && !getStored("kifi_installation_id")) {
```
Comment (2is10) [inline packrat/main.js]: I found the regexp a little confusing. How about `tab.url == webBaseUri() + "/#_=_"`?

Also, it seems the `!getStored("kifi_installation_id")` check could go outside the `api.tabs.on.ready.add` call, as an optimization.

**Comment by 2is10** [comment]
This is much simpler. Thanks. Feel free to merge when you're happy with it.

File: packrat/main.js (hunk 5 above)
Comment (gmethvin) [inline packrat/main.js]: I'm not sure `getStored` can go outside the call. That code is only executed when we load the extension, right? If you authenticate via this handler and we visit kifi.com again, the value will be stale, won't it?

Comment (2is10) [inline packrat/main.js]: You can `api.tabs.on.ready.remove(kifiLoginListener)` upon successful authentication.

Comment (gmethvin) [inline packrat/main.js]: Good idea, thanks.

**Pull request merged**